### ISSUE TITLE
UnnecessaryLet allow for chains

### ISSUE THREAD

**Issue by mkrussel77** [detekt/detekt#4388]
## Expected Behavior

Would like some configuration option on the `UnnecessaryLet` rule to allow using it as a map function at the end of a chain of function calls.

Example:
```
val fontMap = context.resources
         .openRawResource(R.raw.fonts)
         .bufferedReader()
         .use { it.readText() }
         .let { parseFontMap(it) }
``` 

## Current Behavior

This is flag as unnecessary because nothing is null.

## Context

Most warnings from this rule are nice, so I want to keep the rule active, but this is a valid use for `let` that the rule forces us to either not use or suppress.

This pattern is even more useful in cases where you have a function expression body.

**Comment by BraisGabin** [comment]
It is a unnecessary usage of `let` because you can write it like this:

```kotlin
val fontMap = parseFontMap(
    context.resources
        .openRawResource(R.raw.fonts)
        .bufferedReader()
        .use { it.readText() }
)
```

But I agree that this doesn't seems like a really unnecessary let, we could have a flag disables by default but that, if you enable it, it will allow this kind of usages. The problem that I see here is: what is the definition of "chain"? I mean, what should we check to say "this usage is fine". 

```kotlin
val fontMap = parseFonrMap(context.resource)
// vs
val fontMap = context.resource.let { parseFonrMap(it) }
```

```kotlin
val fontMap = parseFonrMap(context.resource.openRawResource(R.raw.fonts))
// vs
val fontMap = context.resource.openRawResource(R.raw.fonts).let { parseFonrMap(it) }
```

**Comment by mkrussel77** [comment]
I don't have a real good answer for how to determine if something is a chain.  I was thinking maybe some property for the number of function calls in the chain.

**Comment by alyssa-glean** [comment]
Not sure what the etiquette is for closed issues, but I wanted to add some support for this. The `UnnecessaryLet` rule seems helpful and did flag some simplifications for us, but it also flagged up chaining scenarios that we would want to keep as they are. Ultimately, we've disabled it for now to avoid having to scatter annotations everywhere this happens.

Here is one of our examples, where there is a chain of calls before using a final `.let`:

```
override fun map(input: Result): Transcript =
        breakIntoChunks(input.words) { it.utterance.contains(".") }
            .map { words -> Sentence(words.map(::Word)) }
            .toList()
            .let(::Transcript)
```

The alternative of wrapping the call chain inside `Transcript` is more complex to read, since you now have to parse the chain but also remember that its nested inside of something. We could also add an extension function along the lines of `fun List<Word>.toTranscript() = Transcript(this)` but that's just boilerplate to get around the problem - the `.let` chain feels like the most idiomatic solution here.

If it's possible to count the chain size, perhaps this could be exposed to the user as a configuration option for the rule? Something like `acceptableChainSize`, and call chains of size >= to this would be ignored by the rule?

**Comment by BraisGabin** [comment]
That have sense, sorry @mkrussel77, I think that I missundertood your issue in the frist place. Should it only be allowed when it is used with a reference method or if there is a lambda too? And how do we define "a chain"? And how much should be the default value?

> Not sure what the etiquette is for closed issues,

This is the correct way to do it. Your comment is related with this so it's good to have the context

**Comment by mkrussel77** [comment]
> That have sense, sorry @mkrussel77, I think that I missundertood your issue in the frist place. Should it only be allowed when it is used with a reference method or if there is a lambda too? And how do we define "a chain"? And how much should be the default value?

Ideally a lambda should be allowed also.  I guess a chain would be calling functions on return values directly without temporary variables.  I think a chain needs at least two calls, but the count should probably be configurable.
